File: codes/python/ch05/higher_order_derivatives.py
#!/usr/bin/env python3
"""
higher_order_derivatives.py

Demonstrates spectral differentiation for higher-order derivatives (up to 4th order)
and compares D² construction methods: matrix squaring vs direct formula.

The test function is u(x) = exp(-sin(2x)) on [0, 2π), which is smooth, periodic,
and has higher frequency content than simpler test functions.

This script generates:
1. A 2×2 figure showing derivatives from 1st to 4th order
2. A comparison figure for D² construction methods
3. A convergence table showing accuracy vs derivative order

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams
from pathlib import Path

# -----------------------------------------------------------------------------
# Publication-quality matplotlib configuration
# -----------------------------------------------------------------------------
rcParams['font.family'] = 'serif'
rcParams['font.serif'] = ['Computer Modern Roman', 'CMU Serif', 'DejaVu Serif']
rcParams['mathtext.fontset'] = 'cm'
rcParams['axes.labelsize'] = 11
rcParams['axes.titlesize'] = 12
rcParams['xtick.labelsize'] = 10
rcParams['ytick.labelsize'] = 10
rcParams['legend.fontsize'] = 10
rcParams['figure.dpi'] = 150
rcParams['savefig.dpi'] = 300
rcParams['text.usetex'] = False
rcParams['axes.linewidth'] = 0.8
rcParams['xtick.major.width'] = 0.8
rcParams['ytick.major.width'] = 0.8

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
N_DEMO = 32  # Grid points for demonstration figure
N_FINE = 500  # Fine grid for plotting exact solutions

# Book color scheme
NAVY = '#142D6E'
SKY = '#7896D2'
CORAL = '#E74C3C'
TEAL = '#16A085'
PURPLE = '#8E44AD'
ORANGE = '#E67E22'

# Output paths
SCRIPT_DIR = Path(__file__).parent
OUTPUT_DIR = SCRIPT_DIR / '..' / '..' / '..' / 'textbook' / 'figures' / 'ch05' / 'python'


# -----------------------------------------------------------------------------
# Test function and its exact derivatives
# -----------------------------------------------------------------------------
def test_function(x):
    """u(x) = exp(-sin(2x))"""
    return np.exp(-np.sin(2 * x))


def exact_derivative_1(x):
    """u'(x) = -2*cos(2x)*exp(-sin(2x))"""
    return -2 * np.cos(2 * x) * np.exp(-np.sin(2 * x))


def exact_derivative_2(x):
    """u''(x) = 4*(sin(2x) + cos²(2x))*exp(-sin(2x))"""
    s = np.sin(2 * x)
    c = np.cos(2 * x)
    return 4 * (s + c**2) * np.exp(-s)


def exact_derivative_3(x):
    """u'''(x) = 8*cos(2x)*(sin²(2x) - 3*sin(2x))*exp(-sin(2x))"""
    s = np.sin(2 * x)
    c = np.cos(2 * x)
    return 8 * c * (s**2 - 3 * s) * np.exp(-s)


def exact_derivative_4(x):
    """u''''(x) computed from the pattern."""
    s = np.sin(2 * x)
    c = np.cos(2 * x)
    # u'''' = 16*exp(-s)*[-s³ + 3s² + 5sc² - 3c² - s²c²]
    # Simplified: factor out and combine terms
    term = -s**3 + 3*s**2 + 5*s*c**2 - 3*c**2 - s**2 * c**2
    return 16 * term * np.exp(-s)


# -----------------------------------------------------------------------------
# Spectral differentiation matrix (periodic)
# -----------------------------------------------------------------------------
def spectral_diff_matrix(N):
    """
    Construct the periodic spectral differentiation matrix.

    Parameters:
        N : int - Number of grid points (should be even)

    Returns:
        D : ndarray (N, N) - Differentiation matrix
        x : ndarray (N,) - Grid points on [0, 2π)
    """
    h = 2 * np.pi / N
    x = h * np.arange(N)
    D = np.zeros((N, N))

    for i in range(N):
        for j in range(N):
            if i != j:
                D[i, j] = 0.5 * ((-1) ** (i - j)) / np.tan((i - j) * np.pi / N)

    return D, x


# -----------------------------------------------------------------------------
# Main visualization functions
# -----------------------------------------------------------------------------
def create_derivatives_figure():
    """Create 2×2 figure showing derivatives from 1st to 4th order."""

    # Construct differentiation matrix
    D, x = spectral_diff_matrix(N_DEMO)

    # Sample test function at grid points
    u = test_function(x)

    # Compute numerical derivatives via matrix powers
    u1_num = D @ u
    u2_num = D @ D @ u
    u3_num = D @ D @ D @ u
    u4_num = D @ D @ D @ D @ u

    # Fine grid for exact solutions
    x_fine = np.linspace(0, 2 * np.pi, N_FINE)
    u1_exact = exact_derivative_1(x_fine)
    u2_exact = exact_derivative_2(x_fine)
    u3_exact = exact_derivative_3(x_fine)
    u4_exact = exact_derivative_4(x_fine)

    # Exact values at grid points for error computation
    u1_exact_grid = exact_derivative_1(x)
    u2_exact_grid = exact_derivative_2(x)
    u3_exact_grid = exact_derivative_3(x)
    u4_exact_grid = exact_derivative_4(x)

    # Compute errors
    err1 = np.max(np.abs(u1_num - u1_exact_grid))
    err2 = np.max(np.abs(u2_num - u2_exact_grid))
    err3 = np.max(np.abs(u3_num - u3_exact_grid))
    err4 = np.max(np.abs(u4_num - u4_exact_grid))

    # Create figure
    fig, axes = plt.subplots(2, 2, figsize=(10, 8))

    derivatives = [
        (u1_exact, u1_num, err1, r"$u'(x)$", "First derivative"),
        (u2_exact, u2_num, err2, r"$u''(x)$", "Second derivative"),
        (u3_exact, u3_num, err3, r"$u'''(x)$", "Third derivative"),
        (u4_exact, u4_num, err4, r"$u''''(x)$", "Fourth derivative"),
    ]

    colors_exact = [NAVY, NAVY, NAVY, NAVY]
    colors_num = [CORAL, TEAL, PURPLE, ORANGE]

    for ax, (exact, num, err, ylabel, title), col_num in zip(
        axes.flat, derivatives, colors_num
    ):
        # Plot exact solution
        ax.plot(x_fine, exact, '-', color=NAVY, linewidth=1.5, label='Exact')
        # Plot numerical solution
        ax.plot(x, num, 'o', color=col_num, markersize=5, label='Spectral')

        ax.set_xlabel(r'$x$')
        ax.set_ylabel(ylabel)
        ax.set_title(title, fontsize=11)

        # Set x-axis ticks
        ax.set_xticks([0, np.pi/2, np.pi, 3*np.pi/2, 2*np.pi])
        ax.set_xticklabels([r'$0$', r'$\pi/2$', r'$\pi$', r'$3\pi/2$', r'$2\pi$'])
        ax.set_xlim(0, 2 * np.pi)

        ax.legend(loc='best', fontsize=9)

        # Add error annotation
        ax.text(0.95, 0.05, f'Max error: {err:.2e}',
                transform=ax.transAxes, fontsize=9,
                verticalalignment='bottom', horizontalalignment='right',
                bbox=dict(boxstyle='round', facecolor='white', alpha=0.8))

        # Clean styling
        ax.spines['top'].set_visible(False)
        ax.spines['right'].set_visible(False)
        ax.spines['left'].set_color('#444444')
        ax.spines['bottom'].set_color('#444444')
        ax.tick_params(colors='#444444')
        ax.yaxis.grid(True, linestyle='-', alpha=0.2, color='#888888')
        ax.xaxis.grid(True, linestyle='-', alpha=0.2, color='#888888')
        ax.set_axisbelow(True)

    fig.suptitle(rf'Higher-Order Spectral Derivatives of $u(x) = e^{{-\sin(2x)}}$ ($N = {N_DEMO}$)',
                 fontsize=12, y=0.98)
    plt.tight_layout(rect=[0, 0, 1, 0.96])

    # Save figure
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    output_file = OUTPUT_DIR / 'higher_order_derivatives.pdf'
    fig.savefig(output_file, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(output_file.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)

    print(f'Figure saved to: {output_file.resolve()}')
    plt.close(fig)

    return err1, err2, err3, err4


def create_d2_comparison_figure():
    """Demonstrate D² via matrix squaring: structure and accuracy."""

    N = 16  # Grid size for visualization

    # Construct D and D²
    D, x = spectral_diff_matrix(N)
    D2 = D @ D

    # Test second derivative accuracy
    u = test_function(x)
    u2_num = D2 @ u
    u2_exact = exact_derivative_2(x)
    error = np.max(np.abs(u2_num - u2_exact))

    # Fine grid for plotting
    x_fine = np.linspace(0, 2 * np.pi, N_FINE)

    # Create figure
    fig, axes = plt.subplots(1, 3, figsize=(12, 4))

    # Panel 1: Structure of D² matrix
    vmax = np.max(np.abs(D2))
    im1 = axes[0].imshow(D2, cmap='RdBu_r', vmin=-vmax, vmax=vmax)
    axes[0].set_title(r'$D^2 = D \cdot D$ (matrix structure)', fontsize=11)
    axes[0].set_xlabel('Column $j$')
    axes[0].set_ylabel('Row $i$')
    plt.colorbar(im1, ax=axes[0], shrink=0.8)

    # Panel 2: Eigenvalue spectrum of D²
    eigvals = np.linalg.eigvals(D2)
    eigvals_sorted = np.sort(eigvals.real)
    # Theoretical eigenvalues for d²/dx²: -k² for k = -N/2+1, ..., N/2
    k_vals = np.arange(-N//2 + 1, N//2 + 1)
    theoretical_eigvals = -k_vals**2

    axes[1].plot(eigvals_sorted, 'o', color=NAVY, markersize=6, label='Numerical')
    axes[1].plot(np.sort(theoretical_eigvals), 's', color=CORAL, markersize=4,
                 markerfacecolor='none', label='Theoretical $-k^2$')
    axes[1].set_xlabel('Index')
    axes[1].set_ylabel('Eigenvalue')
    axes[1].set_title(r'Eigenvalues of $D^2$', fontsize=11)
    axes[1].legend(loc='best', fontsize=9)
    axes[1].grid(True, alpha=0.3)
    axes[1].spines['top'].set_visible(False)
    axes[1].spines['right'].set_visible(False)

    # Panel 3: Second derivative accuracy
    axes[2].plot(x_fine, exact_derivative_2(x_fine), '-', color=NAVY,
                 linewidth=1.5, label='Exact')
    axes[2].plot(x, u2_num, 'o', color=TEAL, markersize=6, label='$D^2 u$')
    axes[2].set_xlabel(r'$x$')
    axes[2].set_ylabel(r"$u''(x)$")
    axes[2].set_title(rf"Second derivative (error = {error:.2e})", fontsize=11)
    axes[2].set_xticks([0, np.pi/2, np.pi, 3*np.pi/2, 2*np.pi])
    axes[2].set_xticklabels([r'$0$', r'$\pi/2$', r'$\pi$', r'$3\pi/2$', r'$2\pi$'])
    axes[2].set_xlim(0, 2 * np.pi)
    axes[2].legend(loc='best', fontsize=9)
    axes[2].grid(True, alpha=0.3)
    axes[2].spines['top'].set_visible(False)
    axes[2].spines['right'].set_visible(False)

    fig.suptitle(rf'Matrix Squaring for Second Derivatives ($N = {N}$)',
                 fontsize=12, y=1.02)
    plt.tight_layout()

    # Save figure
    output_file = OUTPUT_DIR / 'd2_comparison.pdf'
    fig.savefig(output_file, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(output_file.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)

    print(f'Figure saved to: {output_file.resolve()}')
    plt.close(fig)

    return error


def generate_convergence_table():
    """Generate convergence table for different derivative orders."""

    N_values = [8, 16, 32, 64]
    results = []

    print("\n" + "=" * 70)
    print("Convergence Table: Higher-Order Derivatives")
    print(f"Test function: u(x) = exp(-sin(2x))")
    print("=" * 70)
    print(f"{'N':>6} {'Error u\'':>14} {'Error u\'\'':>14} "
          f"{'Error u\'\'\'':>14} {'Error u\'\'\'\'':>14}")
    print("-" * 70)

    for N in N_values:
        D, x = spectral_diff_matrix(N)
        u = test_function(x)

        # Numerical derivatives
        u1_num = D @ u
        u2_num = D @ D @ u
        u3_num = D @ D @ D @ u
        u4_num = D @ D @ D @ D @ u

        # Exact derivatives at grid points
        u1_exact = exact_derivative_1(x)
        u2_exact = exact_derivative_2(x)
        u3_exact = exact_derivative_3(x)
        u4_exact = exact_derivative_4(x)

        # Errors
        err1 = np.max(np.abs(u1_num - u1_exact))
        err2 = np.max(np.abs(u2_num - u2_exact))
        err3 = np.max(np.abs(u3_num - u3_exact))
        err4 = np.max(np.abs(u4_num - u4_exact))

        results.append((N, err1, err2, err3, err4))
        print(f"{N:>6d} {err1:>14.2e} {err2:>14.2e} {err3:>14.2e} {err4:>14.2e}")

    print("=" * 70)

    return results


def main():
    """Main function to generate all figures and tables."""

    print("=" * 70)
    print("Higher-Order Derivatives Demonstration")
    print("=" * 70)

    # Create main figure
    print("\n1. Creating higher-order derivatives figure...")
    err1, err2, err3, err4 = create_derivatives_figure()
    print(f"   Errors: u'={err1:.2e}, u''={err2:.2e}, "
          f"u'''={err3:.2e}, u''''={err4:.2e}")

    # Create D² comparison figure
    print("\n2. Creating D² matrix squaring demonstration...")
    d2_error = create_d2_comparison_figure()
    print(f"   Second derivative error via D²: {d2_error:.2e}")

    # Generate convergence table
    print("\n3. Generating convergence table...")
    results = generate_convergence_table()

    print("\n" + "=" * 70)
    print("All figures generated successfully!")
    print("=" * 70)


if __name__ == '__main__':
    main()

File: codes/julia/ch05/higher_order_derivatives.jl
# higher_order_derivatives.jl
#
# Demonstrates spectral differentiation for higher-order derivatives (up to 4th order)
# and compares D^2 construction methods: matrix squaring vs direct formula.
#
# The test function is u(x) = exp(-sin(2x)) on [0, 2pi), which is smooth, periodic,
# and has higher frequency content than simpler test functions.
#
# This script generates:
# 1. A 2x2 figure showing derivatives from 1st to 4th order
# 2. A comparison figure for D^2 construction methods
# 3. A convergence table showing accuracy vs derivative order
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using Printf

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const N_DEMO = 32   # Grid points for demonstration figure
const N_FINE = 500  # Fine grid for plotting exact solutions

# Book color scheme
const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#8E44AD"
const ORANGE = colorant"#E67E22"

# Output paths
SCRIPT_DIR = @__DIR__
OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch05", "julia")
mkpath(OUTPUT_DIR)

# -----------------------------------------------------------------------------
# Test function and its exact derivatives
# -----------------------------------------------------------------------------

"""u(x) = exp(-sin(2x))"""
test_function(x) = exp(-sin(2x))

"""u'(x) = -2*cos(2x)*exp(-sin(2x))"""
exact_derivative_1(x) = -2 * cos(2x) * exp(-sin(2x))

"""u''(x) = 4*(sin(2x) + cos^2(2x))*exp(-sin(2x))"""
function exact_derivative_2(x)
    s = sin(2x)
    c = cos(2x)
    return 4 * (s + c^2) * exp(-s)
end

"""u'''(x) = 8*cos(2x)*(sin^2(2x) - 3*sin(2x))*exp(-sin(2x))"""
function exact_derivative_3(x)
    s = sin(2x)
    c = cos(2x)
    return 8 * c * (s^2 - 3s) * exp(-s)
end

"""u''''(x) computed from the pattern."""
function exact_derivative_4(x)
    s = sin(2x)
    c = cos(2x)
    term = -s^3 + 3s^2 + 5s * c^2 - 3c^2 - s^2 * c^2
    return 16 * term * exp(-s)
end

# -----------------------------------------------------------------------------
# Spectral differentiation matrix (periodic)
# -----------------------------------------------------------------------------
"""
    spectral_diff_periodic(N)

Construct the periodic spectral differentiation matrix for N equispaced
points on [0, 2pi).
"""
function spectral_diff_periodic(N)
    N < 2 && error("Need at least 2 points")

    h = 2pi / N
    x = h .* (0:N-1)

    D = zeros(N, N)
    for i in 1:N
        for j in 1:N
            if i != j
                diff = i - j
                D[i, j] = 0.5 * ((-1)^diff) / tan(diff * pi / N)
            end
        end
    end

    return D, collect(x)
end

# -----------------------------------------------------------------------------
# Figure 1: 2x2 figure showing derivatives from 1st to 4th order
# -----------------------------------------------------------------------------
function create_derivatives_figure()
    # Construct differentiation matrix
    D, x = spectral_diff_periodic(N_DEMO)

    # Sample test function at grid points
    u = test_function.(x)

    # Compute numerical derivatives via matrix powers
    u1_num = D * u
    u2_num = D * (D * u)
    u3_num = D * (D * (D * u))
    u4_num = D * (D * (D * (D * u)))

    # Fine grid for exact solutions
    x_fine = collect(range(0, 2pi, length = N_FINE))
    u1_exact_fine = exact_derivative_1.(x_fine)
    u2_exact_fine = exact_derivative_2.(x_fine)
    u3_exact_fine = exact_derivative_3.(x_fine)
    u4_exact_fine = exact_derivative_4.(x_fine)

    # Exact values at grid points for error computation
    u1_exact_grid = exact_derivative_1.(x)
    u2_exact_grid = exact_derivative_2.(x)
    u3_exact_grid = exact_derivative_3.(x)
    u4_exact_grid = exact_derivative_4.(x)

    # Compute errors
    err1 = maximum(abs.(u1_num .- u1_exact_grid))
    err2 = maximum(abs.(u2_num .- u2_exact_grid))
    err3 = maximum(abs.(u3_num .- u3_exact_grid))
    err4 = maximum(abs.(u4_num .- u4_exact_grid))

    # Create figure
    fig = Figure(size = (820, 620))

    # Supertitle
    Label(fig[0, 1:2],
          L"Higher-Order Spectral Derivatives of $u(x) = e^{-\sin(2x)}$ ($N = %$(N_DEMO)$)",
          fontsize = 13)

    # Pi tick positions and labels
    pi_ticks = ([0, pi/2, pi, 3pi/2, 2pi],
                [L"0", L"\pi/2", L"\pi", L"3\pi/2", L"2\pi"])

    # Derivative data for each panel
    derivatives = [
        (u1_exact_fine, u1_num, err1, L"u'(x)",    "First derivative"),
        (u2_exact_fine, u2_num, err2, L"u''(x)",   "Second derivative"),
        (u3_exact_fine, u3_num, err3, L"u'''(x)",  "Third derivative"),
        (u4_exact_fine, u4_num, err4, L"u''''(x)", "Fourth derivative"),
    ]

    colors_num = [CORAL, TEAL, PURPLE, ORANGE]

    for (idx, (exact_fine, num, err, ylabel, title)) in enumerate(derivatives)
        row = (idx - 1) ÷ 2 + 1
        col = (idx - 1) % 2 + 1

        ax = Axis(fig[row, col],
                  xlabel = L"x",
                  ylabel = ylabel,
                  title = title,
                  xticks = pi_ticks)

        # Plot exact solution
        lines!(ax, x_fine, exact_fine,
               color = NAVY, linewidth = 1.5, label = "Exact")

        # Plot numerical solution
        scatter!(ax, x, num,
                 color = colors_num[idx], markersize = 5, label = "Spectral")

        xlims!(ax, 0, 2pi)

        axislegend(ax, position = :rt, labelsize = 9)

        # Add error annotation
        text!(ax, 0.95, 0.05,
              text = @sprintf("Max error: %.2e", err),
              fontsize = 9, color = :black,
              space = :relative,
              align = (:right, :bottom),
              )

        # Clean styling
        hidespines!(ax, :t, :r)
        ax.leftspinecolor   = RGBf(0.267, 0.267, 0.267)
        ax.bottomspinecolor = RGBf(0.267, 0.267, 0.267)
        ax.xtickcolor = RGBf(0.267, 0.267, 0.267)
        ax.ytickcolor = RGBf(0.267, 0.267, 0.267)
        ax.ygridvisible = true
        ax.ygridcolor   = RGBAf(0.533, 0.533, 0.533, 0.2)
        ax.ygridstyle   = :solid
        ax.xgridvisible = true
        ax.xgridcolor   = RGBAf(0.533, 0.533, 0.533, 0.2)
        ax.xgridstyle   = :solid
    end

    # Save figure
    output_file = joinpath(OUTPUT_DIR, "higher_order_derivatives.pdf")
    save(output_file, fig)
    save(joinpath(OUTPUT_DIR, "higher_order_derivatives.png"), fig, px_per_unit = 4)
    println("Figure saved to: $output_file")

    return err1, err2, err3, err4
end

# -----------------------------------------------------------------------------
# Figure 2: D^2 matrix squaring demonstration (3-panel)
# -----------------------------------------------------------------------------
function create_d2_comparison_figure()
    N = 16  # Grid size for visualization

    # Construct D and D^2
    D, x = spectral_diff_periodic(N)
    D2 = D * D

    # Test second derivative accuracy
    u = test_function.(x)
    u2_num = D2 * u
    u2_exact_grid = exact_derivative_2.(x)
    error_d2 = maximum(abs.(u2_num .- u2_exact_grid))

    # Fine grid for plotting
    x_fine = collect(range(0, 2pi, length = N_FINE))

    # Create figure
    fig = Figure(size = (960, 340))

    # -------------------------------------------------------------------------
    # Panel 1: Structure of D^2 matrix
    # -------------------------------------------------------------------------
    ax1 = Axis(fig[1, 1],
               xlabel = L"Column $j$",
               ylabel = L"Row $i$",
               title = L"$D^2 = D \cdot D$ (matrix structure)",
               yreversed = true,
               aspect = DataAspect())

    vmax = maximum(abs.(D2))
    hm = heatmap!(ax1, 0:N-1, 0:N-1, D2',
                  colormap = :RdBu,
                  colorrange = (-vmax, vmax))

    Colorbar(fig[1, 1], hm, width = 10, tellwidth = true,
             vertical = true, halign = :right)

    ax1.xticks = [0, N ÷ 2, N - 1]
    ax1.yticks = [0, N ÷ 2, N - 1]

    # -------------------------------------------------------------------------
    # Panel 2: Eigenvalue spectrum of D^2
    # -------------------------------------------------------------------------
    ax2 = Axis(fig[1, 2],
               xlabel = "Index",
               ylabel = "Eigenvalue",
               title = L"Eigenvalues of $D^2$")

    eigvals_d2 = eigvals(D2)
    eigvals_sorted = sort(real.(eigvals_d2))

    # Theoretical eigenvalues for d^2/dx^2: -k^2 for k = -N/2+1, ..., N/2
    k_vals = (-N ÷ 2 + 1):(N ÷ 2)
    theoretical_eigvals = sort(Float64.(-k_vals .^ 2))

    scatter!(ax2, 1:N, eigvals_sorted,
             color = NAVY, markersize = 6, label = "Numerical")
    scatter!(ax2, 1:N, theoretical_eigvals,
             color = CORAL, markersize = 5, marker = :rect,
             strokecolor = CORAL, strokewidth = 1.5,
             label = L"Theoretical $-k^2$")

    axislegend(ax2, position = :lt, labelsize = 9)

    # Clean styling
    hidespines!(ax2, :t, :r)
    ax2.leftspinecolor   = RGBf(0.267, 0.267, 0.267)
    ax2.bottomspinecolor = RGBf(0.267, 0.267, 0.267)
    ax2.ygridvisible = true
    ax2.ygridcolor   = RGBAf(0.533, 0.533, 0.533, 0.3)
    ax2.ygridstyle   = :solid
    ax2.xgridvisible = true
    ax2.xgridcolor   = RGBAf(0.533, 0.533, 0.533, 0.3)
    ax2.xgridstyle   = :solid

    # -------------------------------------------------------------------------
    # Panel 3: Second derivative accuracy
    # -------------------------------------------------------------------------
    pi_ticks = ([0, pi/2, pi, 3pi/2, 2pi],
                [L"0", L"\pi/2", L"\pi", L"3\pi/2", L"2\pi"])

    ax3 = Axis(fig[1, 3],
               xlabel = L"x",
               ylabel = L"u''(x)",
               title = @sprintf("Second derivative (error = %.2e)", error_d2),
               xticks = pi_ticks)

    lines!(ax3, x_fine, exact_derivative_2.(x_fine),
           color = NAVY, linewidth = 1.5, label = "Exact")
    scatter!(ax3, x, u2_num,
             color = TEAL, markersize = 6, label = L"D^2 u")

    xlims!(ax3, 0, 2pi)
    axislegend(ax3, position = :rt, labelsize = 9)

    # Clean styling
    hidespines!(ax3, :t, :r)
    ax3.leftspinecolor   = RGBf(0.267, 0.267, 0.267)
    ax3.bottomspinecolor = RGBf(0.267, 0.267, 0.267)
    ax3.ygridvisible = true
    ax3.ygridcolor   = RGBAf(0.533, 0.533, 0.533, 0.2)
    ax3.ygridstyle   = :solid
    ax3.xgridvisible = true
    ax3.xgridcolor   = RGBAf(0.533, 0.533, 0.533, 0.2)
    ax3.xgridstyle   = :solid

    # Supertitle
    Label(fig[0, 1:3],
          L"Matrix Squaring for Second Derivatives ($N = %$(N)$)",
          fontsize = 13)

    # Save figure
    output_file = joinpath(OUTPUT_DIR, "d2_comparison.pdf")
    save(output_file, fig)
    save(joinpath(OUTPUT_DIR, "d2_comparison.png"), fig, px_per_unit = 4)
    println("Figure saved to: $output_file")

    return error_d2
end

# -----------------------------------------------------------------------------
# Convergence table
# -----------------------------------------------------------------------------
function generate_convergence_table()
    N_values = [8, 16, 32, 64]
    results = Tuple{Int, Float64, Float64, Float64, Float64}[]

    println("\n" * "=" ^ 70)
    println("Convergence Table: Higher-Order Derivatives")
    println("Test function: u(x) = exp(-sin(2x))")
    println("=" ^ 70)
    @printf("%6s %14s %14s %14s %14s\n", "N", "Error u'", "Error u''", "Error u'''", "Error u''''")
    println("-" ^ 70)

    for N in N_values
        D, x = spectral_diff_periodic(N)
        u = test_function.(x)

        # Numerical derivatives
        u1_num = D * u
        u2_num = D * (D * u)
        u3_num = D * (D * (D * u))
        u4_num = D * (D * (D * (D * u)))

        # Exact derivatives at grid points
        u1_exact = exact_derivative_1.(x)
        u2_exact = exact_derivative_2.(x)
        u3_exact = exact_derivative_3.(x)
        u4_exact = exact_derivative_4.(x)

        # Errors
        err1 = maximum(abs.(u1_num .- u1_exact))
        err2 = maximum(abs.(u2_num .- u2_exact))
        err3 = maximum(abs.(u3_num .- u3_exact))
        err4 = maximum(abs.(u4_num .- u4_exact))

        push!(results, (N, err1, err2, err3, err4))
        @printf("%6d %14.2e %14.2e %14.2e %14.2e\n", N, err1, err2, err3, err4)
    end

    println("=" ^ 70)

    return results
end

# -----------------------------------------------------------------------------
# Main function
# -----------------------------------------------------------------------------
function main()
    println("=" ^ 70)
    println("Higher-Order Derivatives Demonstration")
    println("=" ^ 70)

    # Create main figure
    println("\n1. Creating higher-order derivatives figure...")
    err1, err2, err3, err4 = create_derivatives_figure()
    @printf("   Errors: u'=%.2e, u''=%.2e, u'''=%.2e, u''''=%.2e\n",
            err1, err2, err3, err4)

    # Create D^2 comparison figure
    println("\n2. Creating D^2 matrix squaring demonstration...")
    d2_error = create_d2_comparison_figure()
    @printf("   Second derivative error via D^2: %.2e\n", d2_error)

    # Generate convergence table
    println("\n3. Generating convergence table...")
    results = generate_convergence_table()

    println("\n" * "=" ^ 70)
    println("All figures generated successfully!")
    println("=" ^ 70)
end

main()
